pkg/scheduler/algorithm_factory.go

```diff
@@ -273,32 +274,43 @@ func RegisterCustomFitPredicate(policy schedulerapi.PredicatePolicy, args *plugi
 	// generate the predicate function, if a custom type is requested
 	if policy.Argument != nil {
 		if policy.Argument.ServiceAffinity != nil {
+			// map LabelsPresence policy to ConfigProducerArgs that's used to configure the ServiceAffinity plugin.
+			if pluginArgs.ServiceAffinityArgs == nil {
+				pluginArgs.ServiceAffinityArgs = &serviceaffinity.Args{}
+			}
+
+			pluginArgs.ServiceAffinityArgs.Labels = append(pluginArgs.ServiceAffinityArgs.Labels, policy.Argument.ServiceAffinity.Labels...)
+
 			predicateFactory = func(args PluginFactoryArgs) predicates.FitPredicate {
 				predicate, precomputationFunction := predicates.NewServiceAffinityPredicate(
 					args.NodeInfoLister,
 					args.PodLister,
 					args.ServiceLister,
-					policy.Argument.ServiceAffinity.Labels,
+					pluginArgs.ServiceAffinityArgs.Labels,
 				)
 
 				// Once we generate the predicate we should also Register the Precomputation
 				predicates.RegisterPredicateMetadataProducer(policy.Name, precomputationFunction)
 				return predicate
 			}
+			// We use the ServiceAffinity plugin name for all ServiceAffinity custom priorities.
+			// It may get called multiple times but we essentially only register one instance of ServiceAffinity predicate.
+			// This name is then used to find the registered plugin and run the plugin instead of the predicate.
+			name = serviceaffinity.Name
 		} else if policy.Argument.LabelsPresence != nil {
 			// map LabelPresence policy to ConfigProducerArgs that's used to configure the NodeLabel plugin.
-			if args.NodeLabelArgs == nil {
-				args.NodeLabelArgs = &nodelabel.Args{}
+			if pluginArgs.NodeLabelArgs == nil {
+				pluginArgs.NodeLabelArgs = &nodelabel.Args{}
 			}
 			if policy.Argument.LabelsPresence.Presence {
-				args.NodeLabelArgs.PresentLabels = append(args.NodeLabelArgs.PresentLabels, policy.Argument.LabelsPresence.Labels...)
+				pluginArgs.NodeLabelArgs.PresentLabels = append(pluginArgs.NodeLabelArgs.PresentLabels, policy.Argument.LabelsPresence.Labels...)
 			} else {
-				args.NodeLabelArgs.AbsentLabels = append(args.NodeLabelArgs.AbsentLabels, policy.Argument.LabelsPresence.Labels...)
+				pluginArgs.NodeLabelArgs.AbsentLabels = append(pluginArgs.NodeLabelArgs.AbsentLabels, policy.Argument.LabelsPresence.Labels...)
 			}
 			predicateFactory = func(_ PluginFactoryArgs) predicates.FitPredicate {
 				return predicates.NewNodeLabelPredicate(
-					args.NodeLabelArgs.PresentLabels,
-					args.NodeLabelArgs.AbsentLabels,
+					pluginArgs.NodeLabelArgs.PresentLabels,
+					pluginArgs.NodeLabelArgs.AbsentLabels,
 				)
 			}
 			// We use the NodeLabel plugin name for all NodeLabel custom priorities.
```

Comment:
ditto, please move this to the beginning of the if statement. and change lines 323 and 328 to use policyName.

Reply:
done